### DISCUSSION TITLE
Add the ability to use custom classes in @export and Array[Type]

### DISCUSSION THREAD

**Discussion by hidemat** [godotengine/godot-proposals#4918] [Engine Core]
I'm pretty sure there is a reason why this is not allowed yet, but nonetheless, just want to put it out there. This would be nice. 

For example, if I make a `CustomResource` that extends `Resource` the only options that I can think of are:

1. expose it in the editor as `@export var custom : Resource` and then do a cast somewhere in my code `custom = custom as CustomResource`
2. create the `CustomResource` as a built-in type by making a c++ module or GDExtension. The problem with this is that not everybody is comfortable with dealing with c++ modules or GDExtensions, and it feels like overkill in terms of implementing a simple custom resource.

And it's the same situation with the `Array` type. If, for example, I iterate through and Array I need to cast it like so:
```
for element in element_array:
    element = element as CustomClass
    ...
```

I guess I'm ok with these workarounds, but it would be a huge plus for people that are used to static typing. It's a quality-of-life improvement. What do you guys think?

**Comment by Diddykonga** [comment]
This has been a long since discussed topic since 2019.
The PR should hopefully make it for 4.0, but if not then hopefully 4.1

Proposal: https://github.com/godotengine/godot-proposals/issues/18
PR: https://github.com/godotengine/godot/pull/62411

**Reply by hidemat** [reply]
I see. Thanks for replying. I checked the `proposals` but not the `issues`.
